generate.py

```diff
@@ -272,12 +287,30 @@ def main():
     class_name = sys.argv[1]
     url = sys.argv[2]
     host_name = get_host_name(url)
-    testhtml = requests.get(url, headers=HEADERS).content
+    original_class_name = class_name.split("_")[0]
 
-    generate_scraper(class_name, host_name)
-    generate_scraper_test(class_name, host_name, url)
-    generate_test_data(class_name, testhtml)
-    init_scraper(class_name)
+    host_exists = False
+    with open("recipe_scrapers/__init__.py", "r") as source:
+        code = source.read()
+        if original_class_name.lower() in code.lower():
+            host_exists = True
```

Comment:
It's been taking me a while to try to figure out exactly why this is, but I'm uncertain about this part of the code.

One thought: the existing-host matching feels imprecise -- and that's _probably_ fine, but I have a sense that there could be a more robust way to approach this -- or perhaps that it'd be so rare for people to encounter this problem that we could skip handling it.

Roughly speaking: I think it's the treatment of code objects (either from `ast` output -- or in this case, by reading the code itself as a string) as malleable text that feels a bit ambiguous and/or fragile.  Both the lowercase-contains check and also the underscore existence checks/splits.

So I'm trying to weigh up possibly-rarely-encountered-benefit against slight-increase-in-code-fragility, and I'm leaning towards suggesting we not modify this code yet, although perhaps think about it again when we handle `testjson` creation, which we'll want to soon.

(apologies: it's taking me a while to review this.  I still don't think this is the best review possible, and that I'm missing some ideas/explanations)

Reply:
Could an alternative method be to check the SCRAPERS dict?

```python
host_exists = get_host_name(url) in SCRAPERS.keys()
```

Reply:
Possibly... currently `generate.py` is independent of the library (doesn't import anything from it), explaining the approach here.

It feels like it should be possible to exact-match check for an existing item within the parsed syntax tree (using `ast`).  That's easy to say; whether it's so easy to write code to do that, I'm not so sure (and should probably attempt).

Reply:
Backtracking to the use cases here, I think we have a single script interface that is trying to handle two cases:

* Generating a scraper, and creating a corresponding test case.
* Adding a test to an existing scraper.

There are currently two ways to uniquely identify a scraper -- by classname and by filename -- and only one unique identifier for a test case -- by filename.  The test case filename is not currently an argument to `generate.py` for either of the use cases, and that, I think, could be where some of the difficulty handling this situation cleanly originates.

I'm mulling it over a bit more, but I think we might want to consider changing the commandline structure for `generate.py` a little.  Adding an underscore (`_`) to indicate we want to switch into the second use case kinda works.. but it feels a bit like a workaround.

The principle I think I'm attempting to follow there is that a good commandline -- and a composed series of commandline steps -- should be somewhat self-documenting, after expanding any abbreviated flags/switches.

Reply:
Extracting each of the two use cases into a separate verb, and making the test filename unambiguous could result in something like:

```
$ generate.py create_scraper --scraper MyScraper --test-filename myscraper --test-url http://example.org/page1
$ generate.py add_test --scraper MyScraper --test-filename myscraper_2 --test-url http://example.org/page2
```

In the above example, the `--scraper` argument required for both commands.  During `create_scraper` we should check that the scraper _does not_ exist in the `SCRAPERS` dictionary (either by exact-matching for the string in the source code read as a string -- or by exact-matching for a node with that name in the abstract syntax tree).  That logic is reversed for `add_test` -- in that case we need to check that the scraper _does exist_ before adding a test.

The `--test-url` argument is always required for both commands.

The `--test-filename` argument would be optional for `create_scraper` but required for `add_test`.

After all that, I think that renaming `generate.py` to something like `manage_scrapers.py` could make sense (with a small nod to Django).  That'd allow for other commands like `rename_scraper`, `remove_test`, `remove_scraper`, and so on in future -- allowing for higher-level scripting that re-arrange scrapers and/or tests when we need to.

:warning: This is me way overthinking and re-speccing a possible approach.  I'm doing that partly to explain how I'd think through this personally, partly because I'd like to see that kind of approach, and partly to discuss/explore the options and see whether they seem agreeable or not.  In reality, I'm probably OK with the script as-is in this branch, but it'd nag at me and I'd like open a changeset with a few of these ideas (and any other feedback) in future.